### PR TITLE
Saga data backwards compat tests

### PR DESCRIPTION
**Pull request by weralabaj**
Connects to https://github.com/Particular/PlatformDevelopment/issues/737

### CODE REVIEW

**Comment by tmasternak** [comment]
ping @Particular/endtoend-maintainers

**Comment by SimonCropp** [comment]
this PR looks a bit off? diff and commits?

**Comment by tmasternak** [comment]
@SimonCropp what do you mean exactly?

**Comment by SimonCropp** [comment]
sorry looked like it had other commits mixed in. guess not so i will review

File: src/NHPersistenceCompatibilityTests/4.5/4.5.csproj [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    <RootNamespace>Version_4_5</RootNamespace>
+    <AssemblyName>NServiceBus.NHibernate.Tests</AssemblyName>
+    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
+    <FileAlignment>512</FileAlignment>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+    <DebugSymbols>true</DebugSymbols>
+    <DebugType>full</DebugType>
+    <Optimize>false</Optimize>
+    <OutputPath>..\bin\Debug\4.5\</OutputPath>
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/4.5/4.5.csproj]: why is debug path so diff from release?

File: src/NHPersistenceCompatibilityTests/4.5/4.5.csproj [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    </Reference>
+    <Reference Include="NServiceBus.Core, Version=4.7.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\NServiceBus.4.7.12\lib\net40\NServiceBus.Core.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="NServiceBus.NHibernate, Version=4.5.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\NServiceBus.NHibernate.4.5.2\lib\net40\NServiceBus.NHibernate.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="nunit.framework, Version=2.5.10.11092, Culture=neutral, PublicKeyToken=96d09a1eb7f44a77, processorArchitecture=MSIL">
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/4.5/4.5.csproj]: nunit 3?

File: src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+namespace Version_4_5
+{
+    public class NHibernateSessionFactory
+    {
+        public NHibernateSessionFactory()
+        {
+            SessionFactory = new Lazy<ISessionFactory>(Init);
+        }
+
+        private ISessionFactory Init()
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs]: remove private

File: src/NHPersistenceCompatibilityTests/4.5/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/4.5/Properties/AssemblyInfo.cs]: delete this

**Comment by SimonCropp** [comment]
add our R# settings from repo standards

File: src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+using NHibernate.Tool.hbm2ddl;
+using NServiceBus.SagaPersisters.NHibernate.AutoPersistence;
+
+namespace Version_4_5
+{
+    public class NHibernateSessionFactory
+    {
+        public NHibernateSessionFactory()
+        {
+            SessionFactory = new Lazy<ISessionFactory>(Init);
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs]: why lazy?

File: src/NHPersistenceCompatibilityTests/4.5/TestPersistence.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    public void Verify(Guid id, IList<int> ints, string originator)
+    {
+        var unitOfWorkManager = new UnitOfWorkManager { SessionFactory = factory.SessionFactory.Value };
+        var persister = new SagaPersister { UnitOfWorkManager = unitOfWorkManager };
+
+        var data = persister.Get<TestSagaDataWithList>(id);
+
+        Assert.AreEqual(id, data.Id);
+        CollectionAssert.AreEqual(originator, data.Originator);
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/4.5/TestPersistence.cs]: why collection assert?

File: src/NHPersistenceCompatibilityTests/Common/CustomAssemblyLoader.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            }
+
+            public Assembly Resolve(object sender, ResolveEventArgs args)
+            {
+                // Use latest strong name & version when trying to load SDK assemblies
+                var requestedAssembly = new AssemblyName(args.Name);
+                if (requestedAssembly.Name != shortName)
+                    return null;
+
+                Debug.WriteLine("Redirecting assembly load of " + args.Name
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/Common/CustomAssemblyLoader.cs]: perhaps trace if u want this to show up in release mode?

File: src/NHPersistenceCompatibilityTests/Common/LocalPackageResolver.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        public LocalPackageResolver(string localStore)
+        {
+            this.localStore = localStore;
+        }
+
+        public Package GetLocalPackage(PackageInfo packageInfo)
+        {
+            var versionFolder = Path.Combine(localStore, packageInfo.Version);
+            var files = new DirectoryInfo(versionFolder).GetFiles();
+            var package = new Package
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/Common/LocalPackageResolver.cs]: just return directly

**Comment by SimonCropp** [comment]
is there a TC build setup for this?

**Comment by SimonCropp** [comment]
turn on treat warnings as errors

**Comment by SimonCropp** [comment]
there are redirects for fsharp but it doesnt seem to be used

File: src/NHPersistenceCompatibilityTests/Common/AppDomainCreator.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+                AppDomain = appDomain,
+                ProjectAssemblyPath = Path.Combine(startupDir.FullName, package.Info.AssemblyName + ".dll"),
+                PackageVersion = package.Version
+            };
+        }
+
+        void SetupBindingRedirects(Package package, AppDomain appDomain)
+        {
+            var assemblyRedirector = new CustomAssemblyLoader(appDomain);
+            var assemblies = package.Files.Where(f => f.Extension == ".dll").ToList();
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/Common/AppDomainCreator.cs]: if we only care about .dlls why are we loading it with all files?

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+{
+    public abstract class TestRun
+    {
+        private static string BinDirectoryTemplate = "TestRun_{0}_{1}";
+
+        protected AppDomainRunner<T> CreateTestFacade<T>(string versionName)
+        {
+            var packageInfo = new PackageInfo("NServiceBus.NHibernate.Tests", versionName);
+            var package = packageResolver.GetLocalPackage(packageInfo);
+            var appDomainDescriptor = domainCreator.CreateDomain(BinDirectoryTemplate, package);
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs]: why pass BinDirectoryTemplate around? isnt it a constant?

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            var runner = new AppDomainRunner<T>(appDomainDescriptor);
+
+            appDomainDescriptors.Add(appDomainDescriptor);
+
+            return runner;
+        }
+
+        [TestFixtureTearDown]
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs]: TestFixtureTearDown is obsolete

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            appDomainDescriptors = new List<AppDomainDescriptor>();
+            packageResolver = new LocalPackageResolver(path);
+            domainCreator = new AppDomainCreator();
+        }
+
+        LocalPackageResolver packageResolver;
+        AppDomainCreator domainCreator;
+        IList<AppDomainDescriptor> appDomainDescriptors;
+
+        //static string BinDirectorySearchPattern = string.Format(BinDirectoryTemplate, "*", "*");
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs]: delete this

**Comment by SimonCropp** [comment]
seems to be some Newtonsoft redirects. is it even used?

File: src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        using (var session = factory.SessionFactory.Value.OpenSession())
+        {
+            persister.Save(new TestSagaDataWithList
+            {
+                Id = id,
+                OriginalMessageId = id.ToString(),
+                Originator = originator,
+                Ints = data
+            }, new SagaCorrelationProperty("corr", id), new TestSessionProvider(session), new ContextBag())
+                .GetAwaiter()
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs]: looking at this Persist should be async

File: src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+                .GetResult();
+
+            session.Flush();
+        }
+    }
+
+    public void Verify(Guid id, IList<int> ints, string originator)
+    {
+        var session = factory.SessionFactory.Value.OpenSession();
+        var data = persister.Get<TestSagaDataWithList>(id, new TestSessionProvider(session), new ContextBag()).GetAwaiter().GetResult();
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs]: should verify be async?

File: src/NHPersistenceCompatibilityTests/4.5/TestPersistence.cs (hunk 6 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/4.5/TestPersistence.cs]: Fixed. The assert is now done on the `Ints` collection and done in test method.

File: src/NHPersistenceCompatibilityTests/4.5/4.5.csproj (hunk 1 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/4.5/4.5.csproj]: My bad. Fixed.

File: src/NHPersistenceCompatibilityTests/4.5/4.5.csproj (hunk 2 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/4.5/4.5.csproj]: Done.

File: src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs (hunk 3 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs]: Done.

File: src/NHPersistenceCompatibilityTests/4.5/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/4.5/Properties/AssemblyInfo.cs]: Do you mean using or the whole file?

File: src/NHPersistenceCompatibilityTests/Common/CustomAssemblyLoader.cs (hunk 7 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/Common/CustomAssemblyLoader.cs]: Done

File: src/NHPersistenceCompatibilityTests/Common/LocalPackageResolver.cs (hunk 8 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/Common/LocalPackageResolver.cs]: Done.

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs (hunk 11 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs]: Removed

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs (hunk 12 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs]: Done

**Comment by tmasternak** [comment]
@SimonCropp there is no TC build yet. This is the first commit. Regarding redirects I've removed all `app.config` files they were not needed.

/cc: @HEskandari

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs (hunk 10 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/TestRun.cs]: Removed the declaration and in-lined in only place it was used.

File: src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs (hunk 14 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs]: This is called across `AppDomain`s. I did not check that but that probably won't work with correctly.

File: src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs (hunk 13 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs]: This is called across `AppDomain`s. I did not check that but that probably won't work with correctly.

/cc: @HEskandari

File: src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs (hunk 5 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs]: I think it was done to make the usage of `SessionFactory` inside `Persiter` more concise. /cc: @HEskandari

File: src/NHPersistenceCompatibilityTests/4.5/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/4.5/Properties/AssemblyInfo.cs]: Assembly info is metadata for the purpose of deployment. So i assume none of them are required in this sln

**Comment by SimonCropp** [comment]
Can a TC build be created then?

File: src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs (hunk 13 above)
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/7.0/TestPersistence.cs]: Yes u r correct that will not work. Will make do with getawaiter

File: src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs (hunk 5 above)
Comment (HEskandari) [inline src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs]: True. It will be initialized when accessed.

**Comment by HEskandari** [comment]
@SimonCropp I'm on it. Just finishing any remaining comments/feedbacks.

**Pull request merged**